zkevm-circuits/src/evm_circuit/execution/precompiles/ec_mul.rs

```diff
@@ -0,0 +1,519 @@
+use bus_mapping::precompile::{PrecompileAuxData, PrecompileCalls};
+use eth_types::{Field, ToLittleEndian, ToScalar, Word};
+use gadgets::util::{and, not, or, Expr};
+use halo2_proofs::{circuit::Value, plonk::Error};
+
+use crate::{
+    evm_circuit::{
+        execution::ExecutionGadget,
+        step::ExecutionState,
+        util::{
+            common_gadget::RestoreContextGadget,
+            constraint_builder::{ConstrainBuilderCommon, EVMConstraintBuilder},
+            math_gadget::{IsZeroGadget, ModGadget},
+            CachedRegion, Cell, RandomLinearCombination,
+        },
+    },
+    table::CallContextFieldTag,
+    witness::{Block, Call, ExecStep, Transaction},
+};
+
+// Modulus for scalar field Fr
+const FR_N: [u8; 32] = [
+    0x01, 0x00, 0x00, 0xF0, 0x93, 0xF5, 0xE1, 0x43, 0x91, 0x70, 0xB9, 0x79, 0x48, 0xE8, 0x33, 0x28,
+    0x5D, 0x58, 0x81, 0x81, 0xB6, 0x45, 0x50, 0xB8, 0x29, 0xA0, 0x31, 0xE1, 0x72, 0x4E, 0x64, 0x30,
+];
+
+#[derive(Clone, Debug)]
+pub struct EcMulGadget<F> {
+    point_p_x_rlc: Cell<F>,
+    point_p_y_rlc: Cell<F>,
+    scalar_s_rlc: Cell<F>,
+    scalar_s_raw_rlc: Cell<F>,
```

Comment:
you can remove these 2 phase2 cells here. Instead you can simply use `scalar_s: Word<F>` and `scalar_s_raw: Word<F>`:
where `Word<F>` is `RandomLinearCombination<F, 32>`.

Reply:
Could these two cells be retained?... They match the specification in the `PrecompileGadget` and they're checked against the `PaddingGadget`. Removing them makes the input rlc check fail as part of `CALL_OP`.

Reply:
The order in which cells are queried in `PrecompileGadget` depends on the cells allocated in the "next" step, i.e. the `EcMul` step. You can change the cell allocation in `EcMul` and hence accordingly change that order in the `PrecompileGadget`.

You would do this in the `PrecompileGadget` for instance:
```rust
                  let (p_x_rlc, p_y_rlc, scalar_s_raw, scalar_s_modded, r_x_rlc, r_y_rlc) = (
                      cb.query_cell_phase2(),
                      cb.query_cell_phase2(),
                      cb.query_keccak_rlc::<32>(),
                      cb.query_keccak_rlc::<32>(),
                      cb.query_cell_phase2(),
                      cb.query_cell_phase2(),
                  );
                  // ... other code ... //
                  cb.require_equal(
                      "input bytes (RLC) = [ p_x | p_y | s ]",
                      padding_gadget.padded_rlc(),
                      (p_x_rlc.expr() * r_pow_64)
                          + (p_y_rlc.expr() * r_pow_32.expr())
                          + scalar_s_raw.expr(),
                  );
                  // ... other code ... //
```